urldac7fd8c97_vitalymaryin_unsafe_utils_tgz-pJ8-org_vit_unsafeutils_TcpTestJ8/src/org/vit/unsafeutils/tcp/TcpStreamProducer.java

```diff
@@ -0,0 +1,155 @@
+package org.vit.unsafeutils.tcp;
+
+import org.apache.log4j.Logger;
+import org.vit.unsafeutils.api.UnsafeSerializable;
+import org.vit.unsafeutils.serializer.UnsafeSerializer;
+import org.vit.unsafeutils.service.Service;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.nio.channels.*;
+import java.nio.channels.spi.SelectorProvider;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/*
+Copyright 2014 Vitaly Maryin
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+public class TcpStreamProducer extends TcpGenericManager implements Service, Runnable {
+    private static Logger logger = Logger.getLogger(TcpStreamProducer.class);
+    private static boolean debugEnabled = logger.isDebugEnabled();
+    private int selectionOpts = SelectionKey.OP_WRITE;
+
+    @Override
+    public void stop() {
+        started = false;
+        if (serverChannel != null && serverChannel.isOpen())
+            try {
+                serverChannel.close();
+            } catch (IOException e) {
+                logger.error(e.getMessage(), e);
+            }
+    }
+
+    @Override
+    public void start() {
+        try {
+            logger.info("starting up tcp producer ...");
+            selector = initSelector();
+            new Thread(this).start();
+        } catch (IOException e) {
+            logger.error(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public boolean shouldStart() {
+        return true;
+    }
+
+    private Selector initSelector() throws IOException {
+        Selector socketSelector = SelectorProvider.provider().openSelector();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *socketSelector*. The resource is referenced at line 70. The resource is returned at line 72. There are other execution paths that don&#39;t close the resource or return it, for example, when ServerSocketChannel.open throws an exception. To prevent this resource leak, close *socketSelector* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).